### ISSUE TITLE
Rule suggestion: ban all usages of optional chaining

### ISSUE THREAD

**Issue by benny-medflyt** [typescript-eslint/typescript-eslint#1792]
TypeScript 3.7 introduced the "optional chaining" operator `?.`

Lots of people love it, but some projects might want to completely ban it. To quote the wisdom of @everdimension:

>  I don't think that a feature like this belongs in the language.
>
>   - Having a long chain of unknown fields should be an unlikely pattern.
>   - A typed language encourages writing out typed schemas so that you don't get into situations like this.
>     - A schema where each field is unknown to be present doesn't usually make much sense.
>     - ...But those cases where it does make sense the problem should be explicit. But having a syntax to help you hide the unreliability of a schema will encourage creating unreliable schemas. It should be the other way around.
>   - Accessing the data through long chains is likely to be a violation of the Principle of Least Knowledge

<https://github.com/microsoft/TypeScript/issues/30167#issuecomment-468729626>

Also see the following TypeScript issues for trouble that can be caused by optional chaining: <https://github.com/microsoft/TypeScript/issues/34974>, <https://github.com/microsoft/TypeScript/issues/35109>

This rule would be similar to `no-non-null-assertion`, `no-namespace`, and `no-dynamic-delete`

**Comment by jeffvandyke** [comment]
My opinion: This may be a better rule for javascript in general, given that optional chaining isn't a TypeScript-specific language feature, and so would be best prepared for eslint directly or a related plugin.

**Comment by bradzacher** [comment]
I could not disagree more.

> Having a long chain of unknown fields should be an unlikely pattern.
> A typed language encourages writing out typed schemas so that you don't get into situations like this.

That's not true. It entirely depends on your project's context and setup.
Having deep chains of optional objects in a schema is something that can occur pretty commonly.

Two examples that come to mind:
- modelling a DB with 0..1 relationships.
- Graph querying engines, such as GraphQL.

GraphQL is a great example.
When working with a GraphQL API, almost every field is optional. This is by design, because on the server-side, every field is resolved independently of other fields. This is done to isolate each field so that you can query multiple data sources, perform transformations on the data, etc.

If one field throws an error, it culls its subtree. This can potentially happen easily - transient infra errors, bugs, missing data over `0..n` connections, etc. So you have to handle this from a UI.

Due to the graph-like nature of the data, it's completely reasonable to have 3 levels of property access within a single component.
Without optional chaining, this is _incredibly_ painful - especially if you are strictly delineating falsey values from nullish ones.

There's a reason that flow has had support for optional chaining for _years_ - because facebook makes heavy use of this pattern in its front-end codebase.
The graphql queries are backed by Relay, which auto-generates types based on the types presented by the graphql schema.

> Accessing the data through long chains is likely to be a violation of the Principle of Least Knowledge

It's important to note that this feature is not just for long chains, it's also great for short chains - i.e. chains of length one.
The idea is that it's a short-hand syntax - syntactic sugar to make life easier for developers..

```ts
declare const x: null | { a: string };

const y = x === null || x === undefined ? undefined : x.a;
// or
const y = x?.a;
```

---

With all that being said - you can ban it if you want via ESLint's no-restricted-syntax rule.
https://eslint.org/docs/rules/no-restricted-syntax

you can find the nodes you want to ban for your parser via https://astexplorer.net/
For now, those nodes are likely `OptionalMemberExpression` and `OptionalCallExpression`.

**Comment by benny-medflyt** [comment]
For scenarios where your problem domain truly requires you to deal with lots of optional data, then I prefer a more structured approach, using a proxy design pattern, or even better, using an optics framework that has Prisms ([monocle-ts](https://github.com/gcanti/monocle-ts) is good). This will result in a much more robust and maintainable codebase in my opinion, and allows you to do all kinds of powerful and well-defined traversals and transformations of your data. Optional chaining on the other hand, just feels way too ad hoc to me, it blurs logic too much for my comfort, and will surely mask bugs due to its complex semantics.

With regards to the short-hand syntax, it is a matter of taste. Some people will prefer your short version. I personally prefer the longer version because it is more explicit. This is similar to the way some people avoid the short-hand `i++` and instead prefer `i += 1` because it is more explicit.

Of course, most teams are not going to program using optics which is why I think it's fine for people to use optional chaining it if it works for them. But I want to keep optional chaining completely out of my code.

Thank you for the tip to use [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax). It seems that `OptionalCallExpression` is not needed. This rule works for me:

```javascript
        "no-restricted-syntax": [
            "error",
            {
                "selector": "OptionalMemberExpression",
                "message": "Optional Chaining not allowed"
            }
        ],
```
